runtime/compact.c

```diff
@@ -32,6 +32,10 @@
 #include "caml/memprof.h"
 #include "caml/eventlog.h"
 
+#define Caml_gray (1 << 8)
+#define Is_gray_hd(hd) (Color_hd (hd) == Caml_gray)
+#define Grayhd_hd(hd)  (((hd)  & ~Caml_black)  | Caml_gray)
```

Comment:
This is unfortunate and probably needs a discussion around how to best integrate the new compactor from #9728 with this change.

It may be the simplest thing to do for now is rename these definitions and push the rest of the discussion to how compaction will work when more of multicore is merged.

Reply:
In the short time, I'd suggest you leave the "gray"-related macros defined in `<caml/gc.h>`, and just concentrate on not using them in the major GC.  This way the compactor is not affected. 

In the longer term, we should re-think the interface between the compactor and the Multicore OCaml major GC.  My guess is that you'd need an extra pass of sweeping before calling the compactor, so that the compactor only sees "marked" and "unmarked" blocks but no "garbage" blocks.

Reply:
Sure. I've moved them back in to gc.h in the latest commit.

On the second point, I think @kayceesrk had a few ideas on how that might work.

Reply:
In the multicore GC, at the end of the major cycle, the meaning of the GC colours are cycled such that `MARKED -> UNMARKED`, `UNMARKED -> GARBAGE`, `GARBAGE -> MARKED` and `FREE` remains `FREE`. Before cycling, there are no `GARBAGE` objects in the heap since the sweeper has made all of them `FREE`. As a result, just after cycling, there are no `MARKED` objects. Hence, depending on when the compaction is (before or after cycling), one of the colours is available for use in compaction.

Reply:
That should work for compaction but I'm curious to know how you deal with out-of-heap blocks, which need to remain `MARKED` at all times.

Reply:
Multicore uses the bit pattern `NOT_MARKABLE (3 << 8)` for out-of-heap objects. `MARKED`, `UNMARKED` and `GARBAGE` cycle among `0 << 8`, `1 << 8` and `2 << 8`. `FREE` is not a distinguished colour in Multicore (although my comment above made it sound like it is). Sweeping sets the header of `GARBAGE` objects to `0` and adds it to the freelist [1]. More information is in the Multicore wiki [2]. 

[1] https://github.com/ocaml-multicore/ocaml-multicore/blob/172fe18bbc7fcdaa1fc286869c28a2c59ecbf27a/runtime/shared_heap.c#L421-L428. 
[2] https://github.com/ocaml-multicore/ocaml-multicore/wiki/Garbage-collector-invariants#gc-cycles